// react-lazy/lib/js/components/DefaultWrapper.js
'use strict';

exports.__esModule = true;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; } /* eslint-disable prefer-template */


function DefaultWrapper(_ref) {
    var childProps = _ref.childProps,
        children = _ref.children,
        isFailed = _ref.isFailed,
        isLoaded = _ref.isLoaded,
        props = _objectWithoutProperties(_ref, ['childProps', 'children', 'isFailed', 'isLoaded']);

    var className = 'react-lazy-wrapper' + (!isLoaded && !isFailed ? ' react-lazy-wrapper--' + (childProps ? 'loading' : 'placeholder') : '') + (isFailed ? ' react-lazy-wrapper--failed' : '') + (isLoaded ? ' react-lazy-wrapper--loaded' : '') + (props.className ? ' ' + props.className : '');

    return _react2.default.createElement(
        'div',
        Object.assign({}, props, { className: className }),
        children
    );
}

process.env.NODE_ENV !== "production" ? DefaultWrapper.propTypes = {
    children: _propTypes2.default.node,
    childProps: _propTypes2.default.object,
    className: _propTypes2.default.string,
    isFailed: _propTypes2.default.bool,
    isLoaded: _propTypes2.default.bool
} : void 0;

exports.default = DefaultWrapper;
