// react-lazy/lib/js/components/LazyGroup.js
'use strict';

exports.__esModule = true;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _wrap = require('../lib/wrap');

var _DefaultWrapper = require('./DefaultWrapper');

var _DefaultWrapper2 = _interopRequireDefault(_DefaultWrapper);

var _reactIntersectionObserver = require('@researchgate/react-intersection-observer');

var _reactIntersectionObserver2 = _interopRequireDefault(_reactIntersectionObserver);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var LazyGroup = function (_React$PureComponent) {
    _inherits(LazyGroup, _React$PureComponent);

    function LazyGroup(props) {
        _classCallCheck(this, LazyGroup);

        var _this = _possibleConstructorReturn(this, _React$PureComponent.call(this, props));

        _this.loadedImgTags = 0;
        _this.state = { imgTagCount: null, loadedAt: null, viewportAt: null };

        _this.handleViewport = _this.handleViewport.bind(_this);
        _this.handleImgLoaded = _this.handleImgLoaded.bind(_this);
        return _this;
    }

    LazyGroup.prototype.handleImgLoaded = function handleImgLoaded() {
        this.loadedImgTags++;

        if (this.loadedImgTags === this.state.imgTagCount) {
            this.loadedImgTags = 0;
            this.setState({ imgTagCount: null, loadedAt: Date.now() }, this.props.onLoad);
        }
    };

    LazyGroup.prototype.handleViewport = function handleViewport(event, unobserve) {
        var _props = this.props,
            children = _props.children,
            childrenToWrap = _props.childrenToWrap,
            onLoad = _props.onLoad,
            onViewport = _props.onViewport,
            visible = _props.visible;


        if (!event.isIntersecting || !visible) {
            return;
        }

        if (onViewport) {
            onViewport(event);
        }

        if (event.defaultPrevented) {
            return;
        }

        unobserve();

        var imgTagCount = (0, _wrap.countTypesTags)(childrenToWrap, children) || null;
        this.loadedImgTags = 0;
        var viewportAt = Date.now();
        this.setState({ imgTagCount: imgTagCount, loadedAt: !imgTagCount ? viewportAt : null, viewportAt: viewportAt }, !imgTagCount ? onLoad : null);
    };

    LazyGroup.prototype.render = function render() {
        var _props2 = this.props,
            children = _props2.children,
            childrenToWrap = _props2.childrenToWrap,
            childWrapper = _props2.childWrapper,
            clientOnly = _props2.clientOnly,
            component = _props2.component,
            cushion = _props2.cushion,
            ref = _props2.forwardedRef,
            ltIE9 = _props2.ltIE9,
            onLoad = _props2.onLoad,
            onViewport = _props2.onViewport,
            threshold = _props2.threshold,
            viewport = _props2.viewport,
            visible = _props2.visible,
            rest = _objectWithoutProperties(_props2, ['children', 'childrenToWrap', 'childWrapper', 'clientOnly', 'component', 'cushion', 'forwardedRef', 'ltIE9', 'onLoad', 'onViewport', 'threshold', 'viewport', 'visible']);

        var props = Object.assign({ ref: ref }, rest);

        return _react2.default.createElement(
            _reactIntersectionObserver2.default,
            { onChange: this.handleViewport, root: viewport, rootMargin: cushion, threshold: threshold },
            _react2.default.createElement(component, props,
            // swap render once element is visible in viewport
            clientOnly || this.state.viewportAt ? (0, _wrap.wrapTypesToLazyChild)(childrenToWrap, children, childWrapper, this.handleImgLoaded, this.state.viewportAt != null) : (0, _wrap.wrapTypesToNoScript)(childrenToWrap, children, ltIE9, childWrapper))
        );
    };

    return LazyGroup;
}(_react2.default.PureComponent);

LazyGroup.defaultProps = {
    childrenToWrap: ['iframe', 'img'],
    childWrapper: _DefaultWrapper2.default,
    clientOnly: false,
    component: 'div',
    ltIE9: false,
    visible: true
};

process.env.NODE_ENV !== "production" ? LazyGroup.propTypes = {
    clientOnly: _propTypes2.default.bool,
    children: _propTypes2.default.node,
    childrenToWrap: _propTypes2.default.arrayOf(_propTypes2.default.any),
    childWrapper: _propTypes2.default.oneOfType([_propTypes2.default.object, _propTypes2.default.func]),
    component: _propTypes2.default.any,
    cushion: _propTypes2.default.string,
    forwardedRef: _propTypes2.default.any,
    ltIE9: _propTypes2.default.bool,
    onLoad: _propTypes2.default.func,
    onViewport: _propTypes2.default.func,
    threshold: _propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.arrayOf(_propTypes2.default.number)]),
    viewport: _propTypes2.default.oneOfType([_propTypes2.default.string].concat(typeof HTMLElement === 'undefined' ? [] : _propTypes2.default.instanceOf(HTMLElement))),
    visible: _propTypes2.default.bool

    // eslint-disable-next-line react/display-name,react/no-multi-comp
} : void 0;exports.default = _react2.default.forwardRef(function (props, ref) {
    return _react2.default.createElement(LazyGroup, Object.assign({}, props, { forwardedRef: ref }));
});
