// react-lazy/lib/es/components/DefaultWrapper.js
function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

/* eslint-disable prefer-template */
import React from 'react';
import PropTypes from 'prop-types';

function DefaultWrapper(_ref) {
    var childProps = _ref.childProps,
        children = _ref.children,
        isFailed = _ref.isFailed,
        isLoaded = _ref.isLoaded,
        props = _objectWithoutProperties(_ref, ['childProps', 'children', 'isFailed', 'isLoaded']);

    var className = 'react-lazy-wrapper' + (!isLoaded && !isFailed ? ' react-lazy-wrapper--' + (childProps ? 'loading' : 'placeholder') : '') + (isFailed ? ' react-lazy-wrapper--failed' : '') + (isLoaded ? ' react-lazy-wrapper--loaded' : '') + (props.className ? ' ' + props.className : '');

    return React.createElement(
        'div',
        Object.assign({}, props, { className: className }),
        children
    );
}

process.env.NODE_ENV !== "production" ? DefaultWrapper.propTypes = {
    children: PropTypes.node,
    childProps: PropTypes.object,
    className: PropTypes.string,
    isFailed: PropTypes.bool,
    isLoaded: PropTypes.bool
} : void 0;

export default DefaultWrapper;
